src/Common/FileCache.cpp

```diff
@@ -0,0 +1,623 @@
+            file_segments.insert(it, cell->file_segment);
+
+            current_pos = segment_range.right + 1;
+            ++it;
+        }
+
+        if (current_pos <= range.right)
+        {
+            ///   ________]     -- requested range
+            ///   _____]
+            ///        ^
+            /// segmentN
+
+            auto hole_size = range.right - current_pos + 1;
+            auto * cell = addCell(key, current_pos, hole_size, FileSegment::State::EMPTY, cache_lock);
+            file_segments.push_back(cell->file_segment);
+        }
+    }
+
+    /// TODO: remove this extra debug logging.
+    String ranges;
+    for (const auto & s : file_segments)
+        ranges += "\nRange: " + s->range().toString() + ", download state: " + FileSegment::stateToString(s->download_state) + " ";
+    LOG_TEST(log, "Cache getOrSet. Key: {}, range: {}, file_segments number: {}, ranges: {}",
+             keyToStr(key), range.toString(), file_segments.size(), ranges);
+
+    assert(!file_segments.empty());
+    return FileSegmentsHolder(std::move(file_segments));
+}
+
+LRUFileCache::FileSegmentCell * LRUFileCache::addCell(
+    const Key & key, size_t offset, size_t size, FileSegment::State state,
+    std::lock_guard<std::mutex> & /* cache_lock */)
+{
+    /// Create a file segment cell and put it in `files` map by [key][offset].
+
+    if (!size)
+        return nullptr; /// Empty files are not cached.
+
+    if (files[key].contains(offset))
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR,
+            "Cache already exists for key: `{}`, offset: {}, size: {}", keyToStr(key), offset, size);
+
+    auto file_segment = std::make_shared<FileSegment>(offset, size, key, this, state);
+    FileSegmentCell cell(std::move(file_segment), queue);
+
+    auto & offsets = files[key];
+
+    if (offsets.empty())
+    {
+        auto key_path = path(key);
+        if (!fs::exists(key_path))
+            fs::create_directories(key_path);
+    }
+
+    auto [it, inserted] = offsets.insert({offset, std::move(cell)});
+    if (!inserted)
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR,
+            "Failed to insert into cache key: `{}`, offset: {}, size: {}", keyToStr(key), offset, size);
+
+    return &(it->second);
+}
+
+bool LRUFileCache::tryReserve(const Key & key_, size_t offset_, size_t size, std::lock_guard<std::mutex> & cache_lock)
+{
+    auto removed_size = 0;
+    size_t queue_size = queue.size();
+    assert(queue_size <= max_element_size);
+
+    /// Since space reservation is incremental, cache cell already exists if it's state is EMPTY.
+    /// And it cache cell does not exist on startup -- as we first check for space and then add a cell.
+    auto * cell_for_reserve = getCell(key_, offset_, cache_lock);
+
+    /// A cell acquires a LRUQueue iterator on first successful space reservation attempt.
+    if (!cell_for_reserve || !cell_for_reserve->queue_iterator)
```

Comment:
What does it mean `cell_for_server == nullptr`? Looks like in this case we don't need to do anything in this function.

Reply:
> Looks like in this case we don't need to do anything in this function.

No, not true, I added a comment.